src/Capability/State/Internal/Class.hs

```diff
@@ -129,3 +134,39 @@ zoom :: forall innertag t (cs :: [Capability]) inner m a.
 zoom =
   derive @t @'[HasState innertag inner] @cs
 {-# INLINE zoom #-}
+
+--------------------------------------------------------------------------------
+
+data instance Reified tag (HasState tag s) m = ReifiedState
+  { _stateSource :: Reified tag (HasSource tag s) m,
+    _stateSink :: Reified tag (HasSink tag s) m,
+    _state :: forall a. (s -> (a, s)) -> m a
+  }
+
+instance
+  ( Monad m,
+    Reifies s' (Reified tag (HasState tag s) m)
+  ) =>
+  HasSource tag s (Reflected s' (HasState tag s) m)
+  where
+  await_ _ = coerce $ _await $ _stateSource $ reified @s'
+  {-# INLINE await_ #-}
+
+instance
+  ( Monad m,
+    Reifies s' (Reified tag (HasState tag s) m)
+  ) =>
+  HasSink tag s (Reflected s' (HasState tag s) m)
+  where
+  yield_ _ = coerce $ _yield $ _stateSink $ reified @s'
+  {-# INLINE yield_ #-}
```

Comment:
Ah, yes, this makes it a bit annoying to generate with template haskell (even eventually). Maybe template haskell lets us access super-classes and iterate on them.